library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/client/ClientResourceLoaderEvents.java

```diff
@@ -0,0 +1,98 @@
+
+import java.util.Optional;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.resource.ResourceManager;
+
+import org.quiltmc.qsl.base.api.event.Event;
+import org.quiltmc.qsl.resource.loader.api.ResourceLoader;
+import org.quiltmc.qsl.resource.loader.api.reloader.IdentifiableResourceReloader;
+
+/**
+ * Events related to the resource loader of the Minecraft client.
+ */
+@Environment(EnvType.CLIENT)
+public final class ClientResourceLoaderEvents {
+	private ClientResourceLoaderEvents() {
+		throw new UnsupportedOperationException("ClientResourceLoaderEvents only contains static definitions.");
+	}
+
+	/**
+	 * An event indicating the start of the reloading of resource packs on the Minecraft client.
+	 * <p>
+	 * This event should not be used to load resources, use {@link ResourceLoader#registerReloader(IdentifiableResourceReloader)} instead.
+	 */
+	public static final Event<StartResourcePackReload> START_RESOURCE_PACK_RELOAD = Event.create(StartResourcePackReload.class,
+			callbacks -> (client, resourceManager) -> {
+				for (var callback : callbacks) {
+					callback.onStartResourcePackReload(client, resourceManager);
+				}
+			});
+
+	/**
+	 * An event indicating the end of the reloading of resource packs on the Minecraft client.
+	 * <p>
+	 * This event should not be used to load resources, use {@link ResourceLoader#registerReloader(IdentifiableResourceReloader)} instead.
+	 */
+	public static final Event<EndResourcePackReload> END_RESOURCE_PACK_RELOAD = Event.create(EndResourcePackReload.class,
+			callbacks -> (client, resourceManager, error) -> {
+				for (var callback : callbacks) {
+					callback.onEndResourcePackReload(client, resourceManager, error);
+				}
+			});
+
+	/**
+	 * Functional interface to be implemented on callbacks for {@link #START_RESOURCE_PACK_RELOAD}.
+	 *
+	 * @see #START_RESOURCE_PACK_RELOAD
+	 */
+	@FunctionalInterface
+	public interface StartResourcePackReload {
+		/**
+		 * Called before resource packs on the Minecraft client have been reloaded.
+		 *
+		 * @param client          the server
+		 * @param resourceManager the resource manager
+		 */
+		void onStartResourcePackReload(MinecraftClient client, ResourceManager resourceManager);
+	}
+
+	/**
+	 * Functional interface to be implemented on callbacks for {@link #END_RESOURCE_PACK_RELOAD}.
+	 *
+	 * @see #END_RESOURCE_PACK_RELOAD
+	 */
+	@FunctionalInterface
+	public interface EndResourcePackReload {
+		/**
+		 * Called after resource packs on the Minecraft client have been reloaded.
+		 * <p>
+		 * If the reload was not successful, the old resource packs will be kept.
+		 *
+		 * @param client          the client
+		 * @param resourceManager the resource manager
+		 * @param error           present if the resource pack reload failed
+		 */
+		void onEndResourcePackReload(MinecraftClient client, ResourceManager resourceManager, Optional<Throwable> error);
```

Comment:
Why does this take in an `Optional<Throwable>`, while the server equivalent takes in a nullable `Throwable`?

Reply:
This is only mirroring Minecraft's internal code. 
On the client it returns an optional at the end of the resource reloading, and on the server it only returns the bare `Throwable`.
If we absolutely want consistency, either I wrap the error into an Optional on server, or I pass a Nullable error everywhere.

Reply:
Couldn't you just make it return a nullable error everywhere? That would help with consistency, and wouldn't incur any extra cost on the server.